android/tests/java/com/mapswithme/maps/widget/placepage/PlaceOpeningHoursAdapterTest.java

```diff
@@ -0,0 +1,335 @@
+  {
+    // opening_hours = "Mo-Fr 09:00-18:00; Sa 12:00-14:00"
+    Timetable[] timetables = new Timetable[2];
+    timetables[0] = new Timetable(new Timespan(new HoursMinutes(9,0, true),
+                                               new HoursMinutes(18,0, true)),
+                                  new Timespan[0],
+                                  false,
+                                  new int[]{2, 3, 4, 5, 6});
+
+    timetables[1] = new Timetable(new Timespan(new HoursMinutes(12,0, true),
+                                               new HoursMinutes(14,0, true)),
+                                  new Timespan[0],
+                                  false,
+                                  new int[]{7});
+
+    adapter.setTimetables(timetables, Calendar.SUNDAY);
+
+    /* Expected parsed schedule:
+    *  0 - Su, closed
+    *  1 - Mo-Fr, open
+    *  2 - Sa, open
+    * */
+
+    List<WeekScheduleData> schedule = (List<WeekScheduleData>)mWeekScheduleField.get(adapter);
+    assertNotNull(schedule);
+    assertEquals(3, schedule.size());
+    assertTrue(schedule.get(0).isClosed);
+    assertEquals(schedule.get(0).startWeekDay, 1);
+    assertEquals(schedule.get(0).endWeekDay, 1);
+    assertFalse(schedule.get(1).isClosed);
+    assertFalse(schedule.get(2).isClosed);
+  }
+
+  @Test
+  public void test_single_closed_day_monday() throws IllegalAccessException
+  {
+    // opening_hours = "Mo-Fr 09:00-18:00; Sa 12:00-14:00"
+    Timetable[] timetables = new Timetable[2];
+    timetables[0] = new Timetable(new Timespan(new HoursMinutes(9,0, true),
+                                               new HoursMinutes(18,0, true)),
+                                  new Timespan[0],
+                                  false,
+                                  new int[]{2, 3, 4, 5, 6});
+
+    timetables[1] = new Timetable(new Timespan(new HoursMinutes(12,0, true),
+                                               new HoursMinutes(14,0, true)),
+                                  new Timespan[0],
+                                  false,
+                                  new int[]{7});
+
+    adapter.setTimetables(timetables, Calendar.MONDAY);
+
+    /* Expected parsed schedule:
+     *  0 - Mo-Fr, open
+     *  1 - Sa, open
+     *  2 - Su, closed
+     * */
+
+    List<WeekScheduleData> schedule = (List<WeekScheduleData>)mWeekScheduleField.get(adapter);
+    assertNotNull(schedule);
+    assertEquals(3, schedule.size());
+    assertFalse(schedule.get(0).isClosed);  // Mo-Fr - open
+    assertFalse(schedule.get(1).isClosed);  // Sa    - open
+    assertTrue(schedule.get(2).isClosed);   // Su    - closed
+    assertEquals(schedule.get(2).startWeekDay, 1);
+    assertEquals(schedule.get(2).endWeekDay, 1);
+  }
+
+  @Test
+  public void test_multiple_closed_day_monday() throws IllegalAccessException
+  {
+    // opening_hours = "Mo 09:00-18:00; We 10:00-18:00; Fr 11:00-18:00"
```

Comment:
It will make tests very clean and easy to write/read.

Reply:
We do have such function: ` OpeningHours.nativeTimetablesFromString(String)`. But it's implemented in C++ and I'm not sure how to tell tests where to find *.so file. I'll try later. But for now it's easier to parse string by hand.

Reply:
You can add only necessary files and libs from native part:

https://source.android.com/compatibility/tests/development/native-func-e2e

Reply:
Hm, when I try to load `liborganicmaps.so` inside unittests I get `java.lang.UnsatisfiedLinkError`:

```
/.../organicmaps/android/build/intermediates/cmake/webDebug/obj/x86_64/liborganicmaps.so: dlopen(/.../organicmaps/android/build/intermediates/cmake/webDebug/obj/x86_64/liborganicmaps.so, 0x0001): tried: '/.../organicmaps/android/build/intermediates/cmake/webDebug/obj/x86_64/liborganicmaps.so' (not a mach-o file), '/usr/lib/liborganicmaps.so' (no such file)
```

Maybe it's because `liborganicmaps.so` is built for linux and not for MacOS?

Reply:
No. Is it included in the test apk? It should be loaded in the same way as in the main Application class:
```java
  static
  {
    System.loadLibrary("organicmaps");
  }

```

Reply:
`liborganicmaps.so` is included in APK. But APK is running on Android (Linux), while I run tests on Macbook (MacOS). Looks like I'm getting `not a mach-o file` error because of invalid liborganicmaps.so lib format?

Reply:
Wait, how do you run the tests? Aren't they working inside an apk in Emulator on a device?

Reply:
I run unittests with `./gradlew testWebDebugUnitTest`
So it's just a test of Java code on my machine. To run tests on virtual Android Google recommends using [atest](https://source.android.com/compatibility/tests/development/atest).

Reply:
I see. Looks like there is no way to compile and use NDK libraries with local unit tests. Please let us know if you discover one…